linfa-supervised/src/linear_regression/algorithm.rs

```diff
@@ -0,0 +1,119 @@
+    Intercept(Array1<f64>)
+}
+pub struct LinearRegressor {
+    beta : Option<Array1<f64>>,
+    intercept : Intercept,
+}
+*/
+
+/*
+If fit_intercept is false, we suppose that the regression passes throught the origin
+*/
+/*
+The simple linear regression model is
+    y = bX + e  where e ~ N(0, sigma^2 * I)
+In probabilistic terms this corresponds to
+    y - bX ~ N(0, sigma^2 * I)
+    y | X, b ~ N(bX, sigma^2 * I)
+The loss for the model is simply the squared error between the model
+predictions and the true values:
+    Loss = ||y - bX||^2
+The maximum likelihood estimation for the model parameters `beta` can be computed
+in closed form via the normal equation:
+    b = (X^T X)^{-1} X^T y
+where (X^T X)^{-1} X^T is known as the pseudoinverse or Moore-Penrose inverse.
+*/
+pub struct LinearRegression {
+    beta: Option<Array1<f64>>,
+    fit_intercept: bool,
+}
+
+impl LinearRegression {
+    pub fn new(fit_intercept: bool) -> LinearRegression {
+        LinearRegression {
+            beta: None,
+            fit_intercept,
+        }
+    }
+    /* Instead of assert_eq we should probably return a Result, we first have to have a generic error type for all algorithms */
+    pub fn fit<A, B>(&mut self, X: &ArrayBase<A, Ix2>, Y: &ArrayBase<B, Ix1>)
+    where
+        A: Data<Elem = f64>,
+        B: Data<Elem = f64>,
+    {
+        let (n_samples, _) = X.dim();
+
+        // We have to make sure that the dimensions match
+        assert_eq!(Y.dim(), n_samples);
+
+        self.beta = if self.fit_intercept {
+            let dummy_column: Array<f64, _> = Array::ones((n_samples, 1));
+            /*
+                if x is has 2 features and 3 samples
+                x = [[1,2]
+                    ,[3,4]
+                    ,[5,6]]
+                dummy_column = [[1]
+                               ,[1]
+                               ,[1]]
+            */
+            let X_with_ones = stack(Axis(1), &[dummy_column.view(), X.view()]).unwrap();
+            Some(LinearRegression::fit_beta(&X_with_ones, Y))
+        } else {
+            Some(LinearRegression::fit_beta(X, Y))
+        }
+    }
+    fn fit_beta<A, B>(X: &ArrayBase<A, Ix2>, y: &ArrayBase<B, Ix1>) -> Array1<f64>
+    where
+        A: Data<Elem = f64>,
+        B: Data<Elem = f64>,
+    {
+        let rhs = X.t().dot(y);
+        let linear_operator = X.t().dot(X);
+        linear_operator.solve_into(rhs).unwrap()
+    }
+
+    pub fn predict<A>(&self, X: &ArrayBase<A, Ix2>) -> Array1<f64>
+    where
+        A: Data<Elem = f64>,
+    {
+        let (n_samples, _) = X.dim();
+
+        // If we are fitting the intercept, we need an additional column
+        if self.fit_intercept {
+            let dummy_column: Array<f64, _> = Array::ones((n_samples, 1));
+            let X = stack(Axis(1), &[dummy_column.view(), X.view()]).unwrap();
+            match &self.beta {
+                None => panic!("The linear regression estimator has to be fitted first!"),
```

Comment:
This is not ideal - can we refactor the way we build `LinearRegression` to make sure that `beta` is always there when a user calls `predict`?

The easiest way to do this is re-using the same approach I put down in `KMeans`: a structure to hold the hyperparameters of the model (built with/without builder pattern, depending on how many hyperparameters we have there) and a `fit` method on it that returns a fitted `LinearRegression` - basically, using `fit` as constructor. In this way we can be sure that `beta` is there and we have one less `panic` condition 😁